File: Chapter11/C#/SystemBrowser/Binder.cs
using System;
using System.Windows.Forms;
using System.Threading;

namespace SystemBrowser
{
	/// <summary>
	/// This class wires all the top-level objects of the application
	/// </summary>
  public class Binder
  {
    private Builder builder;

    // a handler will be used only during startup, to show progress on the splash screen
    NavigatorFolders.ProgressHandler progressUpdater;

    public Binder(Builder theBuilder)
    {
      builder = theBuilder;
      progressUpdater = new NavigatorFolders.ProgressHandler(builder.formSplash.UpdateProgress);
    }

    public void Bind()
    {
      builder.formMenuToolBar.OnViewFolders += new FormMenuToolBar.UniversalHandler(builder.formMain.ShowFolders);
      builder.formMenuToolBar.OnViewFolders += new FormMenuToolBar.UniversalHandler(builder.userSettings.ShowFolders);

      builder.formMenuToolBar.OnViewSearch += new FormMenuToolBar.UniversalHandler(builder.formMain.ShowSearch);
      builder.formMenuToolBar.OnViewSearch += new FormMenuToolBar.UniversalHandler(builder.userSettings.ShowSearch);

      builder.formMenuToolBar.OnViewIcons += new SystemBrowser.FormMenuToolBar.UniversalHandler(builder.contentFolders.ShowIcons);
      builder.formMenuToolBar.OnViewIcons += new SystemBrowser.FormMenuToolBar.UniversalHandler(builder.contentSearch.ShowIcons);

      builder.formMenuToolBar.OnViewDetails += new SystemBrowser.FormMenuToolBar.UniversalHandler(builder.contentFolders.ShowDetails);
      builder.formMenuToolBar.OnViewDetails += new SystemBrowser.FormMenuToolBar.UniversalHandler(builder.contentSearch.ShowDetails);

      builder.formMenuToolBar.OnUpSelected += new SystemBrowser.FormMenuToolBar.UniversalHandler(builder.navigatorFolders.SelectParentFolder);
      builder.formMenuToolBar.OnAddressChanged += new SystemBrowser.FormMenuToolBar.AddressChangedHandler(builder.navigatorFolders.SelectFolder);

      builder.navigatorFolders.OnFolderChanged += new NavigatorFolders.FolderChangedHandler(builder.contentFolders.Populate);
      builder.navigatorFolders.OnFolderChanged += new NavigatorFolders.FolderChangedHandler(builder.formMenuToolBar.ShowAddress);
      builder.navigatorFolders.OnMessage += new NavigatorFolders.MessageHandler(builder.statusBar.Message);

      builder.contentFolders.OnMessage += new ContentFileList.MessageHandler(builder.statusBar.Message);
      builder.contentFolders.OnFolderDoubleClicked += new SystemBrowser.ContentFileList.FolderDoubleClickedHandler(builder.navigatorFolders.SelectFolder);

      builder.coordinatorSearch.OnSearchRequested += new SystemBrowser.CoordinatorSearch.SearchRequestedHandler(builder.contentSearch.Clear);
      builder.coordinatorSearch.OnSearchStart += new ThreadStart(builder.navigatorSearch.Start);
      builder.coordinatorSearch.OnItemFound += new SystemBrowser.CoordinatorSearch.ItemFoundHandler(builder.contentSearch.Add);
      builder.coordinatorSearch.OnMessage += new SystemBrowser.CoordinatorSearch.MessageHandler(builder.statusBar.Message);

      builder.navigatorSearch.OnSearchRequested += new NavigatorSearch.SearchRequestedHandler(builder.formMenuToolBar.ShowAddress);
      builder.navigatorSearch.OnSearchRequested += new NavigatorSearch.SearchRequestedHandler(builder.coordinatorSearch.StartSearch);
      builder.navigatorSearch.OnItemFound += new NavigatorSearch.ItemFoundHandler(builder.coordinatorSearch.ItemFound);
      builder.navigatorSearch.OnMessage += new NavigatorSearch.MessageHandler(builder.coordinatorSearch.FireMessage);

      builder.contentSearch.OnMessage += new SystemBrowser.ContentSearchResults.MessageHandler(builder.statusBar.Message);
    }

    public void BindFormSplash()
    {
      builder.navigatorFolders.OnProgress += progressUpdater;
    }

    public void UnbindFormSplash()
    {
      builder.navigatorFolders.OnProgress -= progressUpdater;
    }
  }
}

File: Chapter11/VB/SystemBrowser/Binder.vb
' This class wires all the top-level objects of the application

Public Class Binder

  Private _builder As Builder

  ' a handler will be used only during startup, to show progress on the splash screen
  Private _progressUpdater As NavigatorFolders.ProgressHandler

  Public Sub New(ByVal theBuilder As Builder)
    _builder = theBuilder
    _progressUpdater = AddressOf _builder._formSplash.UpdateProgress
  End Sub

  Public Sub Bind()
    AddHandler _builder._formMenuToolBar.OnViewFolders, AddressOf _builder._formMain.ShowFolders
    AddHandler _builder._formMenuToolBar.OnViewFolders, AddressOf _builder._userSettings.ShowFolders

    AddHandler _builder._formMenuToolBar.OnViewSearch, AddressOf _builder._formMain.ShowSearch
    AddHandler _builder._formMenuToolBar.OnViewSearch, AddressOf _builder._userSettings.ShowSearch

    AddHandler _builder._formMenuToolBar.OnViewIcons, AddressOf _builder._contentFolders.ShowIcons
    AddHandler _builder._formMenuToolBar.OnViewIcons, AddressOf _builder._contentSearch.ShowIcons

    AddHandler _builder._formMenuToolBar.OnViewDetails, AddressOf _builder._contentFolders.ShowDetails
    AddHandler _builder._formMenuToolBar.OnViewDetails, AddressOf _builder._contentSearch.ShowDetails

    AddHandler _builder._formMenuToolBar.OnUpSelected, AddressOf _builder._navigatorFolders.SelectParentFolder
    AddHandler _builder._formMenuToolBar.OnAddressChanged, AddressOf _builder._navigatorFolders.SelectFolder

    AddHandler _builder._navigatorFolders.OnFolderChanged, AddressOf _builder._contentFolders.Populate
    AddHandler _builder._navigatorFolders.OnFolderChanged, AddressOf _builder._formMenuToolBar.ShowAddress
    AddHandler _builder._navigatorFolders.OnMessage, AddressOf _builder._statusBar.Message

    AddHandler _builder._contentFolders.OnMessage, AddressOf _builder._statusBar.Message
    AddHandler _builder._contentFolders.OnFolderDoubleClicked, AddressOf _builder._navigatorFolders.SelectFolder

    AddHandler _builder._coordinatorSearch.OnSearchRequested, AddressOf _builder._contentSearch.Clear
    _builder._coordinatorSearch.OnSearchStart = AddressOf _builder._navigatorSearch.Start 'a delegate, not an event
    AddHandler _builder._coordinatorSearch.OnItemFound, AddressOf _builder._contentSearch.Add
    AddHandler _builder._coordinatorSearch.OnMessage, AddressOf _builder._statusBar.Message

    AddHandler _builder._navigatorSearch.OnSearchRequested, AddressOf _builder._formMenuToolBar.ShowAddress
    AddHandler _builder._navigatorSearch.OnSearchRequested, AddressOf _builder._coordinatorSearch.StartSearch
    AddHandler _builder._navigatorSearch.OnItemFound, AddressOf _builder._coordinatorSearch.ItemFound
    AddHandler _builder._navigatorSearch.OnMessage, AddressOf _builder._coordinatorSearch.FireMessage

    AddHandler _builder._contentSearch.OnMessage, AddressOf _builder._statusBar.Message
  End Sub

  Public Sub BindFormSplash()
    AddHandler _builder._navigatorFolders.OnProgress, _progressUpdater
  End Sub

  Public Sub UnbindFormSplash()
    RemoveHandler _builder._navigatorFolders.OnProgress, _progressUpdater
  End Sub

End Class
